File: nodejs/benchmarks-concurrent.js
import { Bench } from "tinybench";
import { chromium } from "playwright";
import { PlaywrightPdfCreator } from "./pdf-creator/playwright-pdf-creator.js";
import { runBench, getFiles } from "./benchmark-utils.js";

export async function runBenchmarksConcurrent() {
  const playwrightPdfCreator = new PlaywrightPdfCreator();
  const playwrightBrowser = await chromium.launch();
  const playwrightBrowserContext = await playwrightBrowser.newContext();

  const nPages = 5;
  const files = await getFiles(nPages);
  const playwrightPages = await createPlaywrightPages(
    playwrightBrowserContext,
    nPages,
  );
  let rotationOffset = 0;

  const bench = new Bench();
  bench
    .add("PlaywrightSerial", async () => {
      for (let i = 0, l = files.length; i < l; i++) {
        await playwrightPdfCreator.fromHtmlWithPage(
          playwrightPages[0],
          files[i],
        );
      }
    })
    .add("PlaywrightConcurrent", async () => {
      rotationOffset++;
      await Promise.all(
        files.map((f, idx) =>
          playwrightPdfCreator.fromHtmlWithPage(
            playwrightPages[(idx + rotationOffset) % nPages],
            f,
          ),
        ),
      );
    });
  await runBench(bench);

  await destroyPlaywrightPages(playwrightPages);
  await playwrightBrowserContext.close();
  await playwrightBrowser.close();
}

async function createPlaywrightPages(browserContext, nPages) {
  const promises = [];
  for (let i = 0; i < nPages; i++) {
    promises.push(browserContext.newPage());
  }
  return await Promise.all(promises);
}

async function destroyPlaywrightPages(pages) {
  await Promise.all(pages.map((p) => p.close()));
}

File: c#/BenchmarksConcurrent.cs
using BenchmarkDotNet.Attributes;
using Microsoft.Playwright;
using PdfBenchmarks.PdfCreator;

namespace PdfBenchmarks
{
    [MemoryDiagnoser]
    public class BenchmarksConcurrent
    {
        private static readonly int nPages = 5;
        private static List<string> ContentValues => Data.GetWikipediaContent(nPages);

        private readonly PlaywrightPdfCreator _playwrightPdfCreator = new();
        private readonly List<IPage> _playwrightPages = [];
        private int rotationOffset = 0;


        [GlobalSetup]
        public async Task GlobalSetup()
        {
            await _playwrightPdfCreator.Initialize();
            await CreatePlaywrightPages();
        }

        [Benchmark]
        public async Task PlaywrightSerial()
        {
            for (int i = 0; i < nPages; i++)
            {
                await _playwrightPdfCreator.FromHtml(ContentValues[i]);
            }
        }

        [Benchmark]
        public async Task PlaywrightConcurrent()
        {
            rotationOffset++;
            var tasks = new List<Task>();
            for (int i = 0; i < nPages; i++)
            {
                tasks.Add(PlaywrightPdfCreator.FromHtml(_playwrightPages[(i + rotationOffset) % nPages], ContentValues[i]));
            }
            await Task.WhenAll(tasks);
        }

        private async Task CreatePlaywrightPages()
        {
            var playwright = await Playwright.CreateAsync();
            var browser = await playwright.Chromium.LaunchAsync();
            var context = await browser.NewContextAsync();
            for (int i = 0; i < nPages; i++)
            {
                _playwrightPages.Add(await context.NewPageAsync());
            }
        }
    }
}